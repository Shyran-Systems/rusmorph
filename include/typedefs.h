#ifndef __TYPEDEFS_H__
#define __TYPEDEFS_H__

typedef unsigned int  word32_t;
typedef unsigned int  word16_t;
typedef unsigned char byte_t;

#endif

