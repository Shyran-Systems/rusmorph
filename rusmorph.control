# rusmorph extension
comment = 'Morphologycal analysers/taggers for Russia'
default_version = '1.0'
module_pathname = '$libdir/rusmorph'
relocatable = true
