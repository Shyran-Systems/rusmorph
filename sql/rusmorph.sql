create extension rusmorph;

set client_encoding to 'utf-8';

select ts_lexize('rusmorph', 'слушать');
select ts_lexize('rusmorph', 'слушаю');
select ts_lexize('rusmorph', 'слушаюсь');
select ts_lexize('rusmorph', 'слушаясь');
select ts_lexize('rusmorph', 'слушаемся');
select ts_lexize('rusmorph', 'слушается');
select ts_lexize('rusmorph', 'слушалось');
select ts_lexize('rusmorph', 'слушались');
select ts_lexize('rusmorph', 'слушалась');
select ts_lexize('rusmorph', 'слушаешься');
select ts_lexize('rusmorph', 'слушаетесь');
select ts_lexize('rusmorph', 'слушаются');
select ts_lexize('rusmorph', 'слушался');
select ts_lexize('rusmorph', 'слушавший');

select ts_lexize('rusmorph', 'чудо');
select ts_lexize('rusmorph', 'чудес');
select ts_lexize('rusmorph', 'чудеса');
select ts_lexize('rusmorph', 'чудесам');
select ts_lexize('rusmorph', 'Чудо');

select ts_lexize('rusmorph', 'Лебедев');
select ts_lexize('rusmorph', 'лебедев');
select ts_lexize('rusmorph', 'Лебедева');
select ts_lexize('rusmorph', 'лебедева');

