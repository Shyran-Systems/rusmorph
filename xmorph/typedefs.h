# if !defined( __typedefs_h__ )
# define __typedefs_h__

# if defined( LIBMORPH_NAMESPACE )
namespace LIBMORPH_NAMESPACE {
# endif  // LIBMORPH_NAMESPACE

// Define common types used in the analyser
# if !defined( __byte_t_defined__ )
#   define  __byte_t_defined__
    typedef unsigned char   byte_t;
# endif  // !__byte_t_defined__

# if !defined( __word16_t_defined__ )
#   define __word16_t_defined__
    typedef unsigned short  word16_t;
# endif  // !__word16_t_defined__

# if !defined( __word32_t_defined__ )
#   define __word32_t_defined__
    typedef unsigned int    word32_t;
# endif  // !__word32_t_defined__

# if defined( LIBMORPH_NAMESPACE )
}
# endif  // LIBMORPH_NAMESPACE

# endif  // __typedefs_h__
