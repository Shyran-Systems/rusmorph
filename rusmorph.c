#include <postgres.h>
#include <tsearch/ts_public.h>
#include <parser/parse_func.h>
#include <mb/pg_wchar.h>
#include "mlma1049.h"
#include "typedefs.h"

PG_MODULE_MAGIC;

PG_FUNCTION_INFO_V1(rusmorph_init);
PG_FUNCTION_INFO_V1(rusmorph_lexize);

size_t client2win(char* dst, size_t dst_size, char const* src, size_t src_size)
{
    PG_TRY();
    {    	
		char* cnv = (char*)pg_do_encoding_conversion((unsigned char*)src, src_size, pg_get_client_encoding(), PG_WIN1251);
		strncpy(dst, cnv, dst_size);
		if (cnv == src && src_size < dst_size) { 
			dst[src_size] = '\0';
			return src_size;
		}
		return strlen(dst);
	}
	PG_CATCH();
    {
		return (size_t)-1;
	}
    PG_END_TRY();
}

size_t win2client(char* dst, size_t dst_size, char const* src, size_t src_size)
{
    PG_TRY();
    {    	
		char* cnv = (char*)pg_do_encoding_conversion((unsigned char*)src, src_size, PG_WIN1251, pg_get_client_encoding());
		strncpy(dst, cnv, dst_size);
		if (cnv == src && src_size < dst_size) { 
			dst[src_size] = '\0';
			return src_size;
		}
		return strlen(dst);
	}
	PG_CATCH();
    {
		return (size_t)-1;
	}
    PG_END_TRY();
}

Datum
rusmorph_init(PG_FUNCTION_ARGS)
{
	PG_RETURN_POINTER(palloc(1));
}

Datum
rusmorph_lexize(PG_FUNCTION_ARGS)
{
	IMlmaMb* pmorph;
	char* in = (char *) PG_GETARG_POINTER(1);
	int			length = PG_GETARG_INT32(2);
	SLemmInfoA  lemmas[0x20];
	char        normal[0x100];
	SGramInfo   agrams[0x40];
	int         i, nlemma;
	TSLexeme   *res;

	if (!length)
        PG_RETURN_POINTER(NULL);  

	mlmaruLoadMbAPI(&pmorph);

	nlemma = pmorph->vtbl->Lemmatize(pmorph, in, length, lemmas, 0x20, normal, 0x100, agrams, 0x40, sfIgnoreCapitals);
	if (nlemma == 0) { 
		PG_RETURN_POINTER(NULL);
	}
	res = palloc0(sizeof(TSLexeme) * (nlemma+1));

	for (i = 0; i < nlemma; i++) 
	{
		res[i].lexeme = pstrdup(lemmas[i].plemma);
	}
	res[i].lexeme = NULL;
	PG_RETURN_POINTER(res);
}
