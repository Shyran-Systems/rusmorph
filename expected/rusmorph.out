create extension rusmorph;
set client_encoding to 'utf-8';
select ts_lexize('rusmorph', 'слушать');
 ts_lexize 
-----------
 {слушать}
(1 row)

select ts_lexize('rusmorph', 'слушаю');
 ts_lexize 
-----------
 {слушать}
(1 row)

select ts_lexize('rusmorph', 'слушаюсь');
 ts_lexize 
-----------
 {слушать}
(1 row)

select ts_lexize('rusmorph', 'слушаясь');
 ts_lexize 
-----------
 {слушать}
(1 row)

select ts_lexize('rusmorph', 'слушаемся');
 ts_lexize 
-----------
 {слушать}
(1 row)

select ts_lexize('rusmorph', 'слушается');
 ts_lexize 
-----------
 {слушать}
(1 row)

select ts_lexize('rusmorph', 'слушалось');
 ts_lexize 
-----------
 {слушать}
(1 row)

select ts_lexize('rusmorph', 'слушались');
 ts_lexize 
-----------
 {слушать}
(1 row)

select ts_lexize('rusmorph', 'слушалась');
 ts_lexize 
-----------
 {слушать}
(1 row)

select ts_lexize('rusmorph', 'слушаешься');
 ts_lexize 
-----------
 {слушать}
(1 row)

select ts_lexize('rusmorph', 'слушаетесь');
 ts_lexize 
-----------
 {слушать}
(1 row)

select ts_lexize('rusmorph', 'слушаются');
 ts_lexize 
-----------
 {слушать}
(1 row)

select ts_lexize('rusmorph', 'слушался');
 ts_lexize 
-----------
 {слушать}
(1 row)

select ts_lexize('rusmorph', 'слушавший');
 ts_lexize 
-----------
 {слушать}
(1 row)

select ts_lexize('rusmorph', 'чудо');
 ts_lexize 
-----------
 {чудо}
(1 row)

select ts_lexize('rusmorph', 'чудес');
   ts_lexize   
---------------
 {чудеса,чудо}
(1 row)

select ts_lexize('rusmorph', 'чудеса');
   ts_lexize   
---------------
 {чудеса,чудо}
(1 row)

select ts_lexize('rusmorph', 'чудесам');
   ts_lexize   
---------------
 {чудеса,чудо}
(1 row)

select ts_lexize('rusmorph', 'Чудо');
 ts_lexize 
-----------
 {чудо}
(1 row)

select ts_lexize('rusmorph', 'Лебедев');
 ts_lexize 
-----------
 {Лебедев}
(1 row)

select ts_lexize('rusmorph', 'лебедев');
 ts_lexize 
-----------
 {Лебедев}
(1 row)

select ts_lexize('rusmorph', 'Лебедева');
 ts_lexize 
-----------
 {Лебедев}
(1 row)

select ts_lexize('rusmorph', 'лебедева');
 ts_lexize 
-----------
 {Лебедев}
(1 row)

